### DISCUSSION TITLE
Please provide detailed instructions for FUZZ of AARCH64 binary/so in Frida mode

### DISCUSSION THREAD

**Discussion by YeJZ** [AFLplusplus/AFLplusplus#1070] [Q&A]
Hello, I am trying to use AFL's Frida mode to FUZZ unsourced binaries. At present, I have successfully executed the make command in the frida-mode directory, but when I use "make ARCH=ARM64",I got this:

collect2: error: ld returned 1 exit status
GNUmakefile:171: recipe for target '/home/XXXXX/AFLplusplus/AFLplusplus/frida_mode/build/afl-frida-trace.so' failed
make: *** [/home/xxxxxx/AFLplusplus/AFLplusplus/frida_mode/build/afl-frida-trace.so] Error 1

I don't know what command I should use to test AARCH64 binaries or SO files .Please help me ,thank you !

**Comment by andreafioraldi** [comment]
Looks like you are not on an arm64 machine...

**Comment by vanhauser-thc** [comment]
@YeJZ you cannot fuzz a different architecture with frida. so x64 => x64, or aarch64 => aarch64.
to fuzz a different architecture you have to use qemu_mode

**Comment by YeJZ** [comment]
> @YeJZ you cannot fuzz a different architecture with frida. so x64 => x64, or aarch64 => aarch64.
> to fuzz a different architecture you have to use qemu_mode

Thanks for the answer, assuming I am on an AARCH64 device, do I need to port the source code of AFL++ to the AARCH64 device to compile and execute "make frida"?

**Comment by YeJZ** [comment]
> Looks like you are not on an arm64 machine...

Yes, I compiled AFL++ on an x86 PC, and tried to use Frida mode to Fuzz the unsourced binary on the Anroid device (AARCH64). The reason for not using the QEMU mode is that the  data object construction is very complicated, so I want to use an in-process passive fuzz method like Frida-Fuzzer.

**Comment by hexcoder-** [comment]
> > @YeJZ you cannot fuzz a different architecture with frida. so x64 => x64, or aarch64 => aarch64.
> > to fuzz a different architecture you have to use qemu_mode
> 
> Thanks for the answer, assuming I am on an AARCH64 device, do I need to port the source code of AFL++ to the AARCH64 device to compile and execute "make frida"?

No, it already has been ported for AARCH64-Linux, so no porting work necessary from your side.
I checked it on a RaspberryPi with Manjaro-Linux.

**Comment by YeJZ** [comment]
> > > @YeJZ you cannot fuzz a different architecture with frida. so x64 => x64, or aarch64 => aarch64.
> > > to fuzz a different architecture you have to use qemu_mode
> > 
> > 
> > Thanks for the answer, assuming I am on an AARCH64 device, do I need to port the source code of AFL++ to the AARCH64 device to compile and execute "make frida"?
> 
> No, it already has been ported for AARCH64-Linux, so no porting work necessary from your side.
> I checked it on a RaspberryPi with Manjaro-Linux.

Thank you for the answer. Now I downloaded the source code of AFL++ and need to test the binary running on the Android device using AFL-Frida mode. Based on your experience, should I execute AFL++ compilation commands on X86? Can you briefly describe the process of running AFL++ on Raspberry Pi? thank you very much

**Comment by vanhauser-thc** [comment]
@hexcoder-

**Comment by hexcoder-** [comment]
Sorry, I have no experience with fuzzing on Android nor do I use frida myself.

As for the Raspberry Pi 4, I have two of them, one with Raspberry Pi OS (32-bit) and one with Manjaro Linux (64 bit).
All I did was installing the necessary tool chain to build AFL++ on it, build AFL++ and let the tests run.
This was exactly like on a PC (since it is Linux on both platforms). 

What you want to know is probably first hand experiences on fuzzing binaries on Android, so I would suggest asking @joeyjiaojg 
 or @domenukk or ask in the discord channel.

**Comment by domenukk** [comment]
Sorry, no good input from my side. Optionally, try libafl frida mode, cross compiling that for android is well tested https://github.com/AFLplusplus/LibAFL/tree/main/fuzzers/frida_libpng

**Comment by YeJZ** [comment]
> Sorry, I have no experience with fuzzing on Android nor do I use frida myself.
> 
> As for the Raspberry Pi 4, I have two of them, one with Raspberry Pi OS (32-bit) and one with Manjaro Linux (64 bit).
> All I did was installing the necessary tool chain to build AFL++ on it, build AFL++ and let the tests run.
> This was exactly like on a PC (since it is Linux on both platforms).
> 
> What you want to know is probably first hand experiences on fuzzing binaries on Android, so I would suggest asking @joeyjiaojg
> or @domenukk or ask in the discord channel.

Thank you for your answer.
And yes, what I want to know is the  first hand experiences on fuzzing binaries on Android.

**Comment by YeJZ** [comment]
> Sorry, no good input from my side. Optionally, try libafl frida mode, cross compiling that for android is well tested https://github.com/AFLplusplus/LibAFL/tree/main/fuzzers/frida_libpng

Thank you for your answer,I will try it out according to your suggestion.
